File: scenegraph/src/main/cpp/trackscenenode.cpp
/******************************************************************************
 *
 * CoreWall / Corelyzer - An Initial Core Description Tool
 * Copyright (C) 2004, 2005 Arun Gangadhar Gudur Rao, Julian Yu-Chung Chen,
 * Sangyoon Lee, Electronic Visualization Laboratory, University of Illinois 
 * at Chicago
 *
 * This software is free software; you can redistribute it and/or modify it 
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either Version 2.1 of the License, or 
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 * License for more details.
 * 
 * You should have received a copy of the GNU Lesser Public License along
 * with this software; if not, write to the Free Software Foundation, Inc., 
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * Questions or comments about CoreWall should be directed to 
 * cavern@evl.uic.edu
 *
 *****************************************************************************/
 
#include "trackscenenode.h"
#include "textureresource_ex.h"
#include "graph.h"
#include "freedraw.h"
#include "canvas.h"

//================================================================
void update_track_dimensions(TrackSceneNode* t);

//================================================================
void create_track(const char* sessionName, const char* trackName, TrackSceneNode*& ptr)
{
    TrackSceneNode* t = new TrackSceneNode();

    // Session name
    t->sessionName = new char[strlen(sessionName) + 1];
    strcpy(t->sessionName, sessionName);

    // Track name
    t->name = new char[strlen(trackName) + 1];
    strcpy(t->name, trackName);

    t->highlight = false;
    t->show = true;   
    t->selectedSection = -1;
    t->modelvec.clear();
    t->zorder.clear();
    t->px = 0;
    t->py = 0;
    t->w  = 0;
    t->h  = 0;
	t->nextPos = 0.0f;
	t->movable = false; // only in depth direction

    t->highlight_color = new GLfloat[3];
    t->highlight_color[0] = 1.0f;
    t->highlight_color[1] = 1.0f;
    t->highlight_color[2] = 0.0f;

    ptr = t;
}

//================================================================
int append_model( TrackSceneNode* t, CoreSection* c)
{
    if( !t || !c ) return -1;
    int pos = -1;
    
    for( int i = 0; i < t->modelvec.size(); ++i)
    {
        if( t->modelvec[i] == NULL )
        {
            t->modelvec[i] = c;
            
            // make sure we don't have more than one i in zorder
            if( t->zorder.size() == 0)
            {
                t->zorder.push_back(i);
                return i;
            }
            
            bool found = false;
            for( int k = 0; k < t->zorder.size(); ++k)
            {
                if( t->zorder[k] == i )
                {
                    if(!found)
                        found = true;
                    else
                        t->zorder[k] = -1;
                }
            }

            if(!found)
            {
                t->zorder.push_back(i);
                bring_model_front(t,i);
            }

            update_track_dimensions(t);
            return i;
        }
    }

    t->modelvec.push_back(c);
    t->zorder.push_back( t->modelvec.size() - 1);
    update_track_dimensions(t);

    return t->modelvec.size() - 1;
}

//================================================================
void free_track(TrackSceneNode* t)
{
    if( t )
    {
        free_all_models(t);
        delete t;
    }
}

//================================================================
void free_all_models(TrackSceneNode* t)
{
    if(!t) return;
    for( int i = 0; i < t->modelvec.size(); ++i)
    {
        free_section_model(t->modelvec[i]);
        t->modelvec[i] = NULL;
    }

    t->modelvec.clear();
    t->zorder.clear();

    delete [] t->sessionName;
    delete [] t->name;
    delete [] t->highlight_color;
}

//================================================================
void free_model(TrackSceneNode* t, int section)
{
    if(!is_section_model(t,section)) return;
    free_section_model( t->modelvec[section]);
    t->modelvec[section] = NULL;
}

//================================================================
bool is_section_model(TrackSceneNode* t, int section)
{
    if(!t) return false;
    if(section < 0 || section > t->modelvec.size() - 1) return false;
    return (t->modelvec[section] != NULL);
}

//================================================================
void highlight_model(TrackSceneNode* t, int section, bool highlight)
{
    if(!is_section_model(t,section)) return;
    t->modelvec[section]->highlight = highlight;
}

void setShow(TrackSceneNode* t, bool isShow)
{
    if(t)
    {
        t->show = isShow;
    }
}

//================================================================
void bring_model_front(TrackSceneNode* t, int section)
{
    if(!is_section_model(t,section)) return;
    
    // find it in the zorder
   
    int i;
    int z = -1;
    for( i = 0; i < t->zorder.size(); ++i)
    {
        if( t->zorder[i] == section )
        {
            z = i;
            i = t->zorder.size();
        }
    }

    // never found!!!

    if( z < 0 )
    {
        printf("NEVER FOUND SECTION %d\n", section);
        t->zorder.push_back(section);
        return;
    }

    for( i = z; i > 0; --i)
        t->zorder[i] = t->zorder[i - 1];

    t->zorder[0] = section;
}

//================================================================
void render_track(TrackSceneNode* t, Canvas* c)
{
#ifdef DEBUG
    printf("\n\t--- Render Track ---\n");
#endif

    if( !t || !c ) 
    {
#ifdef DEBUG
        printf("Track or Canvas is NULL!\n");
#endif
        return;
    }

    if(!t->show) return;

    // move our canvas
    float x, y, z;

#ifdef DEBUG
    get_camera_position(c->camera, &x, &y, &z);
    printf("Track b4: Camera position %.2f %.2f\n", x, y);
    printf("Track translation: %.2f %.2f\n", -t->px, -t->py);
#endif

    // orig: translate_camera( c->camera, -t->px, -t->py, 0);
    if(get_horizontal_depth()) {
        translate_camera( c->camera, -t->px, -t->py, 0);
    } else {
        translate_camera( c->camera, t->py, -t->px, 0);
    }

    glTranslatef( t->px, t->py, 0);

#ifdef DEBUG
    get_camera_position( c->camera, &x, &y, &z);
    printf("Track after: Camera position %.2f, %.2f\n", x, y);
    printf("Track has zorder of size %d\n", t->zorder.size());
#endif

    int i;
    // draw back to front
    glPushMatrix();
    for( i = t->zorder.size() - 1; i > - 1; --i)
    {
#ifdef DEBUG
        printf("Rendering Core Section %d\n", t->zorder[i]);
#endif

        if( t->zorder[i] >= 0 && t->modelvec[ t->zorder[i] ] != NULL)
            render_section_model( t->modelvec[ t->zorder[i] ], c);
    }
    glPopMatrix();

    // draw plugin free draw rectangles, scale so x,y,w,h are in meters
    glPushMatrix();
    float scale = 1.0 / c->dpi_x * CM_PER_INCH / 100.0f;
    float indep_scale = c->w / c->w0;
    glScalef(1.0 /scale,1.0 /scale,1.0 /scale);
    for( i = 0; i < t->freedrawvec.size(); i++)
    {
        if( is_free_draw_scale_independent(t->freedrawvec[i]))
        {
            glScalef( indep_scale, indep_scale, 1.0f);
            render_free_draw(t->freedrawvec[i]);
            glScalef( 1.0 / indep_scale, 1.0 / indep_scale, 1.0f);
        }
        else
        {
            render_free_draw(t->freedrawvec[i]);

        }
    }

    glPopMatrix();

    glTranslatef( -t->px, -t->py, 0);

    // orig: translate_camera( c->camera, t->px, t->py, 0);
    if(get_horizontal_depth()) {
        translate_camera( c->camera, t->px, t->py, 0);
    } else {
        translate_camera( c->camera, -t->py, t->px, 0);
    }
}

//================================================================
CoreSection* get_track_section(TrackSceneNode* t, int section)
{
    if(!is_section_model(t,section)) return NULL;
    return t->modelvec[section];
}

//================================================================
int get_track_section_zorder_length(TrackSceneNode* t)
{
    if(!t) return 0;
    return t->zorder.size();
}

//================================================================
void get_track_section_zorder(TrackSceneNode* t, int* order)
{
    if(!t || !order) return;
    for( int i = 0; i < t->zorder.size(); ++i)
    {
        order[i] = t->zorder[i];
    }
}

//================================================================
void update_track_dimensions(TrackSceneNode* t)
{
    if(!t) return;
    float max_x = 0.0, max_y = 0.0;
    float min_x = 0.0, min_y = 0.0;

    for( int i = 0; i < t->zorder.size(); ++i)
    {
        CoreSection* c = t->modelvec[ t->zorder[i] ];
        if( !c ) continue;
        
        if( c->px < min_x ) min_x = c->px;
        if( c->py < min_y ) min_y = c->py;
        
        float cw, ch;
        float gw, gh;

        cw = get_texset_src_width( c->src );
        ch = get_texset_src_height( c->src );
        
        if(c->orientation == PORTRAIT)
        {
            float t;
            t  = cw;
            cw = ch;
            ch = t;
        }

//        printf("Checking image of dims %f x %f, with dpis of %f, %f\n",
//               cw,ch,c->dpi_x,c->dpi_y);

        if( cw > 0)
        {
            // scale to describe number of inches actually covered
            cw /= c->dpi_x;
            if( c->px + cw > max_x) max_x = (c->px + cw);
        }

        if( ch > 0)
        {
            // scale to describe number of inches actually covered
            ch /= c->dpi_y;
            if( c->py + cw > max_y) max_y = (c->py + ch);
        }

        for( int i = 0; i < c->graphvec.size(); i++)
        {
            Box* b = get_graph_box( c, c->graphvec[i] );
            if(!b) continue;
            delete b;
        }
    }

    t->w = max_x - min_x;
    t->h = max_y - min_y;

//    printf("Updated track dimensions %f x %f\n", t->w, t->h);
}

//================================================================
void attach_free_draw_to_track(TrackSceneNode* t, int fdid)
{
    if(!t) return;
    if(!is_free_draw_rectangle(fdid)) return;
    for( int i = 0; i < t->freedrawvec.size(); i++)
    {
        if( t->freedrawvec[i] == -1)
        {
            t->freedrawvec[i] = fdid;
            return;
        }
    }

    t->freedrawvec.push_back(fdid);

}

//================================================================
void detach_free_draw_from_track(TrackSceneNode* t, int fdid)
{
    if(!t) return;
    if(!is_free_draw_rectangle(fdid)) return;

    for( int i = 0; i < t->freedrawvec.size(); i++) 
    {
        if( t->freedrawvec[i] == fdid)
        {
            t->freedrawvec[i] = -1;
            return;
        }
    }
}

//================================================================
void push_section_to_end(TrackSceneNode* t, int section)
{
    if(!is_section_model(t,section)) return;
    float dpix, dpiy;
    get_canvas_dpi( 0, &dpix, &dpiy );

    float max_x = 0.0f;
    CoreSection* max_sec = NULL;

    for( int i = 0; i < t->modelvec.size(); i++)
    {
        if(!is_section_model(t,i)) continue;
        if( i == section ) continue;

        CoreSection* cs = get_track_section( t, i);
        if( cs->px >= max_x )
        {
            max_x = cs->px;
            max_sec = cs;
        }
    }

#ifdef DEBUG
    printf("Max x: %f, Max Sec %d\n", max_x, max_sec);
#endif

    if( max_sec == NULL ) return;

    CoreSection* sec = get_track_section( t, section );

    // do we have image in max_sec?
	if (max_sec->src != -1)
	{
		// we have section image
		sec->px = max_sec->px + 
			(get_texset_src_width(max_sec->src) * (dpix / max_sec->dpi_x));
	}
	else
	{
		// in case that the last one is graph only section
		// then, use section width instead of texture width
		sec->px = max_sec->px + max_sec->width * INCH_PER_CM * dpix;
	}
	
	// update section depth var
	sec->depth = sec->px * CM_PER_INCH / dpix;

#ifdef DEBUG
    printf("Final position for pushing to end %f\n", sec->px);
#endif

}

void set_track_highlight_color(TrackSceneNode* t, float r, float g, float b)
{
    if(!t) return;

    t->highlight_color[0] = r;
    t->highlight_color[1] = g;
    t->highlight_color[2] = b;
}

File: app/src/main/java/corelyzer/data/TrackSceneNode.java
/******************************************************************************
 *
 * CoreWall / Corelyzer - An Initial Core Description Tool
 * Copyright (C) 2004, 2005 Arun Gangadhar Gudur Rao, Julian Yu-Chung Chen
 * Electronic Visualization Laboratory, University of Illinois at Chicago
 *
 * This software is free software; you can redistribute it and/or modify it 
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either Version 2.1 of the License, or 
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public 
 * License for more details.
 * 
 * You should have received a copy of the GNU Lesser Public License along
 * with this software; if not, write to the Free Software Foundation, Inc., 
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * Questions or comments about CoreWall should be directed to 
 * cavern@evl.uic.edu
 *
 *****************************************************************************/

package corelyzer.data;

import java.util.Enumeration;
import java.util.Hashtable;
import java.util.Vector;

import corelyzer.ui.AlphanumComparator;

/**
 * Data structure to hold core track information in Java side.
 * 
 * 
 */
public class TrackSceneNode implements Comparable<TrackSceneNode> {
	/** Name of the track */
	String trackName;
	String DISId = "N/A_N/A";

	float topDepth = 0.0f;
	float MCDDepth = 0.0f;
	float length = 0.0f;

	/** ID of the track */
	int trackId;

	AnnotationRepository annorepo;

	/** Core sections in this track */
	Vector<CoreSection> secVec;
	/** Z order of Core sections in this track */
	Vector<Integer> zOrder;

	Boolean highlight;
	CoreSectionImage[] selectedSection;

	/** hashtable for section name (key) and coresection (value) */
	Hashtable secnameHash;
	/** hashtable for section id (key) and coresection (value) */
	Hashtable secidHash;

	/**
	 * Constructor
	 * 
	 * @param id
	 *            track ID from the native scene graph
	 */
	public TrackSceneNode(final int id) {
		highlight = false;
		// this.imgVec = new Vector<corelyzer.data.CoreSectionImage>();
		this.secVec = new Vector<CoreSection>();
		this.zOrder = new Vector<Integer>();
		trackId = id;
		secnameHash = new Hashtable();
		secidHash = new Hashtable();
	}

	/**
	 * Constructor
	 * 
	 * @param n
	 *            track name
	 * @param id
	 *            track ID from the native scene graph
	 */
	public TrackSceneNode(final String n, final int id) {
		this(id);
		this.setName(n);
	}

	/** Append a object to this node as a child in scene graph */
	public void addChild(final Object child, final int secId, final int gid) {
		if (child == null) {
			return;
		}

		Boolean b1 = false;
		Boolean b2 = false;
		Boolean b3 = false;

		try {
			b1 = Class.forName("corelyzer.data.CoreSectionImage").isInstance(child);
			b2 = Class.forName("corelyzer.data.AnnotationThread").isInstance(child);
			b3 = Class.forName("corelyzer.data.CoreSectionGraph").isInstance(child);
		} catch (ClassNotFoundException e) {
			System.err.println("--- ClassNotFoundException!");
		}

		if (b1) {
			// System.out.println("[DEBUG] child is corelyzer.data.CoreSectionImage");
			this.addCoreSectionImage((CoreSectionImage) child, secId, gid);
		} else if (b2) {
			// System.out.println("[DEBUG] child is corelyzer.data.AnnotationThread");
			this.annorepo.moveToFront(annorepo.insert((AnnotationThread) child));
		} else if (b3) {
			// System.out.println("[DEBUG] child is corelyzer.data.Graph");
			this.addCoreSectionGraph((CoreSectionGraph) child, secId, gid);
		} else {
			System.err.println("[DEBUG] Error in corelyzer.data.TrackSceneNode#addChild()");
		}
	}

	// Add CoreSection cs to the end of secVec
	public void addCoreSection(final CoreSection cs) {
		addCoreSection(cs, -1);
	}
	
	// If insertIndex != -1, insert CoreSection cs into secVec at insertIndex, otherwise add to the end 
	public void addCoreSection(final CoreSection cs, final int insertIndex)
	{
		if (cs == null) {
			return;
		}

		// bail out if a section with this name already exists
		CoreSection sec = this.getCoreSection(cs.getName());
		if (sec != null) {
			return;
		} else {
			if ( insertIndex == -1 )
				this.secVec.addElement(cs);
			else
				this.secVec.insertElementAt(cs, insertIndex);
			
			this.zOrder.addElement(this.secVec.size() - 1);
			this.secnameHash.put(cs.getName(), cs);
			this.secidHash.put(cs.getId(), cs);
		}
	}
	
	// graph, section gid, graph gid
	public void addCoreSectionGraph(final CoreSectionGraph grp, final int secId, final int gid) {

		if (grp == null) {
			return;
		}
		// assume that we already created section before this point
		CoreSection sec = this.getCoreSection(grp.getName());
		if (sec != null) {
			sec.addGraph(grp);
		}
	}

	void addCoreSectionImage(final CoreSectionImage img, final int secId, final int gid) {

		if (img == null) {
			return;
		}
		// assume that we already created section before this point
		CoreSection sec = this.getCoreSection(img.getSectionName());
		if (sec != null) {
			sec.setImage(img);
		}
	}

	public boolean containsSectionName(final String name) {
		if (secVec.isEmpty()) {
			return false;
		}

		return this.secnameHash.containsKey(name);
	}

	/** Access method */
	AnnotationRepository getAnnoTrreadRepo() {
		return this.annorepo;
	}

	public CoreSection getCoreSection(final int i) {
		if (i < 0 || i >= secVec.size()) {
			return null;
		}

		return this.secVec.elementAt(i);
	}

	public CoreSection getCoreSection(final String secname) {

		if (secVec.isEmpty() || secname == null) {
			return null;
		}

		for (CoreSection cs : this.secVec) {
			// So DIS' SS_XXXX image section can match for MSCL data sections
			// ImageSection example: SS_313_27_A_9_2_1
			// MSCL data section: 313_27_A_9_2_1
			if (cs.getName().endsWith(secname)) {
				return cs;
			}
		}

		return null;
		// return (CoreSection) this.secnameHash.get(secname);
	}

	public CoreSection getCoreSectionByGID(final int i) {

		if (secVec.isEmpty()) {
			return null;
		}

		return (CoreSection) this.secidHash.get(i);
	}

	/** Access method */
	CoreSectionGraph getCoreSectionGraph(final int i) {
		return this.secVec.elementAt(i).getCoreSectionGraph(i);
	}

	/** Access method */
	CoreSectionImage getCoreSectionImage(final int i) {
		return this.secVec.elementAt(i).getCoreSectionImage();
	}

	int getCoreSectionIndex(final CoreSection sec) {
		if (sec == null || this.secVec.isEmpty()) {
			return -1;
		}
		return this.secVec.indexOf(sec);
	}

	public Vector<CoreSection> getCoreSections() {
		return secVec;
	}

	/** Access method */
	Vector<Integer> getCoreZOrder() {
		return this.zOrder;
	}

	public String getDISId() {
		return DISId;
	}

	/** Access method */
	public int getId() {
		return this.trackId;
	}

	public float getLength() {
		return length;
	}

	public float getMCDDepth() {
		return MCDDepth;
	}

	/** Access method */
	public String getName() {
		return this.trackName;
	}

	/** Access method */
	public int getNumCores() {
		return this.secVec.size();
	}

	/** Access method */
	CoreSectionImage[] getSelected() {
		return this.selectedSection;
	}

	public float getTopDepth() {
		return topDepth;
	}

	public void removeAllCoreSection() {

		// remove graph in the section first
		CoreSection sec = null;
		Enumeration en = this.secVec.elements();
		while (en.hasMoreElements()) {
			sec = (CoreSection) en.nextElement();
			sec.removeAllGraph();
		}

		// clean up section & zOrder vector
		this.secnameHash.clear();
		this.secidHash.clear();
		this.secVec.clear();
		this.zOrder.clear();
	}

	public void removeCoreSection(final CoreSection sec) {

		if (sec == null) {
			return;
		}

		// remove graph including native resource
		sec.removeAllGraph();
		this.secnameHash.remove(sec.getName());
		this.secidHash.remove(sec.getId());
		int idx = secVec.indexOf(sec);
		this.secVec.removeElement(sec);
		this.zOrder.removeElement(idx);
	}

	/** Access method */
	void setDimensions(final float w, final float h) {
	}

	public void setDISId(final String DISId) {
		this.DISId = DISId;
	}

	/** Access method */
	void setId(final int i) {
		this.trackId = i;
	}

	public void setLength(final float length) {
		this.length = length;
	}

	public void setMCDDepth(final float MCDDepth) {
		this.MCDDepth = MCDDepth;
	}

	/** Access method */
	public void setName(final String name) {
		this.trackName = name;
	}

	/** Access method */
	void setSectionDPI(final CoreSectionImage img, final float dpi) {
	}

	public void setTopDepth(final float topDepth) {
		this.topDepth = topDepth;
	}

	/**
	 * Override method to return the name of the track as the object
	 * 
	 * @return track name
	 */

	@Override
	public String toString() {
		return this.trackName;
	}

	public void Update() {
	}
	
	public int compareTo(TrackSceneNode track) {
		return AlphanumComparator.compare(this.getName(), track.getName());
	}
}
